lib/metrics.js

```diff
@@ -0,0 +1,161 @@
+const { randomUUID } = require('crypto');
+const {
+  Registry,
+  Counter,
+  Histogram,
+  Gauge,
+  collectDefaultMetrics,
+} = require('prom-client');
+
+const { cpu, mem, netstat } = osu = require('node-os-utils');
+
+module.exports = function configure(config) {
+  // eslint-disable-next-line max-len
+  // See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md
+  const labels = {
+    'faas_name': config.name || 'anonymous',
+    'faas_id': config.id || process.env.FAAS_ID || `${process.env.POD_NAME}-${randomUUID()}`,
+    'faas_instance': config.instance || process.env.POD_NAME,
+    'faas_runtime': config.runtime || 'Node.js',
+  };
+
+  // Create a metrics register and add our identifying labels to it
+  const register = new Registry();
+  register.setDefaultLabels(labels);
+  collectDefaultMetrics({ register });
+
+  // invoked on GET to /metrics
+  async function callMetrics(req, res) {
+    try {
+      res.headers('Content-Type', register.contentType);
+      res.send(await register.metrics());
+    } catch (ex) {
+      res.code(500).send(ex);
+    }
+  }
+
+  const invocations = new Counter({
+    name: 'faas_invocations',
+    help: 'The number of times this function was invoked',
+    registers: [register],
+  });
+
+  const execLatency = new Histogram({
+    name: 'faas_execution_latency',
+    help: 'The time it took for this function to be invoked and return',
+    registers: [register],
+  });
+
+  const errors = new Counter({
+    name: 'faas_errors',
+    help: 'The number of function invocation failures',
+    registers: [register],
+  });
+
+  const coldStartLatency = new Histogram({
+    name: 'faas_cold_start_latency',
+    help: 'The time it took for this function to be in the ready state after the container started',
+    registers: [register],
+  });
+
+  const queueLatency = new Histogram({
+    name: 'faas_queue_latency',
+    help: 'The time spent in a system queue before the function was invoked',
+    registers: [register],
+  });
+
+  new Gauge({
+    name: 'faas_cpu_utilization',
+    help: 'The CPU utilization of this function instance',
+    collect() {
+      cpu.usage()
+        .then(usage => {
+          if (supported(usage)) {
+            this.set(usage);
+          }})
+        .catch(console.error);
+    },
+    registers: [register],
+  });
+
+  new Gauge({
+    name: 'faas_mem_utilization',
+    help: 'The memory utilization of this function instance',
+    collect() {
+      mem.used()
```

Comment:
Since we might have multiple invocations within the same container what we probably want here is the allocation per invocation. Is this the case here when requests are in flight in parallel?